### ISSUE TITLE
Serialization project outline

### ISSUE THREAD

**Issue by h-mayorquin** [SpikeInterface/spikeinterface#1235]
Hi @alejoe91 @samuelgarcia 
I want to propose here an order to work in a series of serialization issues, discuss and get your feedback and direction. I would like to work on this in the following order, each item being a tailored PR to keep the workflow modular and specific.  Let me know what you think of the order and priorities.

- [x] Adding serialization capabilities to recorders and sorters  #1227. 
- [x] Using the serialization capabilites to add initial instantiation capabilties to graph computation pipeline (#1209) that Sam developed #1271
- [x] I: Using the serialization capabilites to remove `to_dict` mechanism in the sorting components module #1272
- [x] II: Using the serialization capabilites to remove `to_dict` mechanism in the sorting components module #1325
- [x] III: Using the serialization capabilites to remove `to_dict` mechanism in the pre-processing module #1446
- [x] IV: Using the serialization capabilites to remove `to_dict` mechanism in the postprocessing module #1455
- [x] V: Using the serialization capabilites to remove `to_dict` mechanism in the curation module #1456
- [x] Using the same capabilities to avoid random and computation in preprocessing init class (solving issue #1118).
- [x] Added a JSON decoder to have a more standard structure #1397.
- [x] Separated flags for whether an extractor objects is dumpable to json and pickable https://github.com/SpikeInterface/spikeinterface/pull/1775
- [x] Separate serialization and json-nification infraestructure.

What do you think? If you agree with this proposal I would like to request a review of #1227 so I can start working on the other items.

**Comment by alejoe91** [comment]
Hi Ramon!

It sounds like a good plan to me and I totally endorse it. @samuelgarcia what do you think?

**Comment by samuelgarcia** [comment]
Cool.
Lets discuss thsi with a call soon.
Remove the to_dict in parralel processing is good.

Here some task are independant. I would make separate small PRs for fast discussion and merging.

**Comment by h-mayorquin** [comment]
Yes, I am thinking on making a PR per point above (or maybe more for the later items). Let's agree on when to do a call at the end of today's meeting : ).

**Comment by h-mayorquin** [comment]
Relevant for this project:
https://github.com/SpikeInterface/spikeinterface/pull/1831

**Comment by alejoe91** [comment]
@h-mayorquin can we close this? I checked the last box after #2027

The last serialization issue (https://github.com/SpikeInterface/spikeinterface/issues/1118) is kind of standalone

**Comment by h-mayorquin** [comment]
Yes, I agree, and it was also wrongly scoped. I have been reading the pickling protocol section in the docs and I suspect we might be able to pass numpy buffers without serializing at all but that will be another issue!

I think the top comment looks very nice with all of those to-do checked in : D
